### PR TITLE
Arcus.Security.Core

### PR DESCRIPTION
**Pull request by SamVanhoutte**
Created `ISecretProvider` and `KeyNotFoundException`

### CODE REVIEW

File: src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Arcus.Security.Core.Exceptions
+{
+    /// <summary>
+    /// Exception, thrown when no secret was found, using the given name.
+    /// </summary>
+    public class SecretNotFoundException : Exception
```
Comment (stijnmoreels) [inline src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs]: Good to have custom Exceptions but they should be marked as `[Serializable]` to conform with the guidelines

File: src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public string Name { get;  }
+        public SecretNotFoundException(string name) : base($"The secret {name} was not found.")
```
Comment (stijnmoreels) [inline src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs]: Could you assert for non-blank `string` values and maybe add another ctor to pass along the inner exception?
Thanks

File: src/Arcus.Security.Core/ISecretProvider.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Threading.Tasks;
+using Arcus.Security.Core.Exceptions;
+
+namespace Arcus.Security.Core
+{
+    /// <summary>
+    /// <see cref="ISecretProvider"/> allows developers to build specific Secret key providers.
+    /// </summary>
+    public interface ISecretProvider
```
Comment (stijnmoreels) [inline src/Arcus.Security.Core/ISecretProvider.cs]: Nice clean simple docs. I like it!

File: src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+using System.Text;
+
+namespace Arcus.Security.Core.Exceptions
+{
+    /// <summary>
+    /// Exception, thrown when no secret was found, using the given name.
+    /// </summary>
+    public class SecretNotFoundException : Exception
+    {
+        public string Name { get;  }
```
Comment (stijnmoreels) [inline src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs]: I should look into the formatting guidelines of the Arcus project, but I think there's something about having a blank line between properties and members + having them below the ctor(s).

Not sure though; not blocking for me.

File: src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs (hunk 1 above)
Comment (SamVanhoutte) [reply src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs]: Agree !  Fixed with new commit.

File: src/Arcus.Security.Core/ISecretProvider.cs (hunk 3 above)
Comment (SamVanhoutte) [reply src/Arcus.Security.Core/ISecretProvider.cs]: thanks

File: src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs (hunk 4 above)
Comment (SamVanhoutte) [reply src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs]: I changed it according to your request, but I'll leave it up to you and @tomkerkhove to detail it out.

File: src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs (hunk 2 above)
Comment (SamVanhoutte) [reply src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs]: did it by leveraging Guard.Net (which is our recommended nuget package for validation) - also added innerexception & unit test for this

Comment (stijnmoreels) [reply src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs]: :+1:

File: src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Arcus.Security.Core.Exceptions
+{
+    /// <summary>
+    /// Exception, thrown when no secret was found, using the given name.
+    /// </summary>
+    [Serializable]
+    public class SecretNotFoundException : Exception
+    {
+        public SecretNotFoundException(string name) : this(name, null)
```
Comment (stijnmoreels) [inline src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs]: Ok, last suggestion: the public docs for the ctor of this type could be valuable but this again is probably writen down in the guidelines.

File: src/Arcus.Security.Core/ISecretProvider.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Threading.Tasks;
+using Arcus.Security.Core.Exceptions;
+
+namespace Arcus.Security.Core
```
Comment (tomkerkhove) [inline src/Arcus.Security.Core/ISecretProvider.cs]: Please move this to an interface folder + update namespace

File: src/Arcus.Security.Tests.Unit/Core/SecretNotFoundExceptionTests.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Arcus.Security.Core.Exceptions;
+using Xunit;
+
+namespace Arcus.Security.Tests.Unit.Core
+{
+    public class SecretNotFoundExceptionTests
```
Comment (tomkerkhove) [inline src/Arcus.Security.Tests.Unit/Core/SecretNotFoundExceptionTests.cs]: Please add happy path

File: src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        /// <param name="innerException">Inner exception that can be passed to base exception</param>
+        /// <exception cref="ArgumentException">The name must not be empty</exception>
+        /// <exception cref="ArgumentNullException">The name must not be null</exception>
+        public SecretNotFoundException(string name, Exception innerException) : base($"The secret {name} was not found.", innerException)
+        {
+            Guard.NotNullOrEmpty(name, nameof(name));
+            Name = name;
+        }
+
+        public string Name { get; }
```
Comment (tomkerkhove) [inline src/Arcus.Security.Core/Exceptions/SecretNotFoundException.cs]: Please add xml docs

File: src/Arcus.Security.Core/ISecretProvider.cs [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        /// <summary>
+        /// Retrieves the secret value, based on the given name
+        /// </summary>
+        /// <param name="name">The name of the secret key</param>
+        /// <returns>Returns a <see cref="Task{string}"/> that contains the secret key</returns>
+        /// <exception cref="ArgumentException">The name must not be empty</exception>
+        /// <exception cref="ArgumentNullException">The name must not be null</exception>
+        /// <exception cref="SecretNotFoundException">The secret was not found, using the given name</exception>
+        Task<string> GetAsync(string name);
```
Comment (tomkerkhove) [inline src/Arcus.Security.Core/ISecretProvider.cs]: We don't suffix with async in Arcus, sorry for bad issue description

Comment (SamVanhoutte) [reply src/Arcus.Security.Core/ISecretProvider.cs]: Hehe, first thing I though, when seeing the description "Really?" :)

**Pull request merged**